### PR TITLE
feat(displacement): add setDisplacement functionality

### PR DESCRIPTION
**Pull request by RichardLindhout**
Work in progress, will work on this in my spare time and I plan to have this finished before 15th of december.

Update: seems almost finished need to some testing on iOS and Android.

### CODE REVIEW

File: android/rctmgl/src/main/java/com/mapbox/rctmgl/location/LocationManager.java [hunk 1]
```diff
@@ -76,11 +82,21 @@ public void removeLocationListener(OnUserLocationChange listener) {
...
     public void enable() {
         if (!PermissionsManager.areLocationPermissionsGranted(context)) {
             return;
         }
+
+        // remove existing listeners
+        locationEngine.removeLocationUpdates(this);
+
+        // refresh location engine request with new values
+        this.buildEngineRequest();
```
Comment (ferdicus) [inline android/rctmgl/src/main/java/com/mapbox/rctmgl/location/LocationManager.java]: ah, nice - didn't event think about rebuilding the LocationEngineRequest!

**Comment by ferdicus** [comment]
Hey there, yeah this is way better! We could even think of exposing the other settings on the JS side as well (something for a later PR I guess :)).

On suggestion though: Maybe split this PR into 1) JS lint updates 2) setSmallestDisplacement feature?
This PR includes a lot of changes that have nothing to do with 2).

File: android/rctmgl/local.properties [hunk 2]
```diff
@@ -0,0 +1,8 @@
+## This file must *NOT* be checked into Version Control Systems,
+# as it contains information specific to your local configuration.
+#
+# Location of the SDK. This is only used by Gradle.
+# For customization when using a Version Control System, please read the
+# header note.
+#Tue Nov 26 20:39:01 CET 2019
+sdk.dir=/Users/richardlindhout/Library/Android/sdk
```
Comment (ferdicus) [inline android/rctmgl/local.properties]: 😞

File: ios/RCTMGL/RCTMGLLocationManager.m [hunk 3]
```diff
@@ -54,10 +54,18 @@ - (void)start
...
+        [locationManager setDistanceFilter:(minDisplacement)];
         isListening = YES;
     });
 }
 
+- (void)setMinDisplacement:(CLLocationDistance)minDisplacement
+{
+     dispatch_async(dispatch_get_main_queue(), ^{
+        [locationManager setDistanceFilter:(minDisplacement)];
+    });
```
Comment (ferdicus) [inline ios/RCTMGL/RCTMGLLocationManager.m]: wow, Objective-C sure looks pretty - what does the `^` do?

File: javascript/components/UserLocation.js [hunk 4]
```diff
@@ -58,13 +58,15 @@ class UserLocation extends React.Component {
...
     onUpdate: PropTypes.func,
+    minDisplacement: PropTypes.number,
 
     children: PropTypes.any,
   };
 
   static defaultProps = {
     animated: true,
     visible: true,
+    minDisplacement: 0.1, // 10 cm
```
Comment (ferdicus) [inline javascript/components/UserLocation.js]: hey, I think a `minDisplacement` of 10cm won't have any changing effect, because the GPS position will jump that amount anyways. 
I think we should simply set it to `0` per default - anything under 5m doesn't do much I found out.
Disclaimer: I've only tested on an Android device.

File: ios/RCTMGL.xcodeproj/xcshareddata/xcschemes/RCTMGL.xcscheme [hunk 5]
```diff
@@ -0,0 +1,67 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kristfal) [inline ios/RCTMGL.xcodeproj/xcshareddata/xcschemes/RCTMGL.xcscheme]: Is this file needed?

File: android/rctmgl/src/main/java/com/mapbox/rctmgl/modules/RCTMGLLocationModule.java [hunk 6]
```diff
@@ -69,12 +70,27 @@ public String getName() {
...
+    public void start(float minDisplacement) {
         isEnabled = true;
+        mMinDisplacement = minDisplacement;
         startLocationManager();
     }
 
+    @ReactMethod
+    public void setMinDisplacement(float minDisplacement) {
+        mMinDisplacement = minDisplacement;
+        if (isEnabled) {
```
Comment (kristfal) [inline android/rctmgl/src/main/java/com/mapbox/rctmgl/modules/RCTMGLLocationModule.java]: I think we should only continue here if the new value is different than the old.

**Comment by kristfal** [comment]
Good job. 

Added some initial comments. Also please note that iOS build is failing on CI.

Let me know when you want a full review :)

File: ios/RCTMGL/RCTMGLLocationManager.m (hunk 3 above)
Comment (RichardLindhout) [reply ios/RCTMGL/RCTMGLLocationManager.m]: I have no idea ;-) it's the first time I'm programming objective c and I'm mostly copying stuff from other places and hope for the best ;(

File: javascript/components/UserLocation.js (hunk 4 above)
Comment (RichardLindhout) [reply javascript/components/UserLocation.js]: Agreed! I hope the GPS in my device is more reliable since 5m is too much for me I think.

File: ios/RCTMGL.xcodeproj/xcshareddata/xcschemes/RCTMGL.xcscheme (hunk 5 above)
Comment (RichardLindhout) [reply ios/RCTMGL.xcodeproj/xcshareddata/xcschemes/RCTMGL.xcscheme]: No however I think it should be added to the gitignore then which I will do!

File: android/rctmgl/local.properties (hunk 2 above)
Comment (RichardLindhout) [reply android/rctmgl/local.properties]: No however I think it should be added to the gitignore then which I will do!

**Comment by RichardLindhout** [comment]
Thanks, will do that when it's ready!

File: ios/RCTMGL.xcodeproj/xcshareddata/xcschemes/RCTMGL.xcscheme (hunk 5 above)
Comment (ferdicus) [reply ios/RCTMGL.xcodeproj/xcshareddata/xcschemes/RCTMGL.xcscheme]: maybe just use `react-natives` up to date [`gitignore`](https://github.com/facebook/react-native/blob/master/.gitignore) plus the mapbox additions

File: javascript/components/UserLocation.js (hunk 4 above)
Comment (ferdicus) [reply javascript/components/UserLocation.js]: you clearly work with better GPS than I 😅

**Comment by ferdicus** [comment]
@kristfal, do we have a recent contributors list? 
I don't even know who the resident iOS/ android person is 😅

**Comment by kristfal** [comment]
@ferdicus I don't have anything better than https://github.com/react-native-mapbox-gl/maps/graphs/contributors 

Also github is quite sensible in recommended reviewers.

**Comment by ferdicus** [comment]
@RichardLindhout : let me know if you need any help with this?
I think I need additional rights from your side to actually touch anything - would be super happy to move this along.

Cheers

**Comment by RichardLindhout** [comment]
@ferdicus I gave you rights, please only edit the patch-6 branch ;)

**Comment by ferdicus** [comment]
> @ferdicus I gave you rights, please only edit the patch-6 branch ;)

Sure thing, no worries.

I will remove all js lint related stuff and would open a completely separate PR if you're ok with that.

**Comment by RichardLindhout** [comment]
@ferdicus The problem is when I save something it does automatically since the prettier config has changed I think.

File: android/rctmgl/src/main/java/com/mapbox/rctmgl/modules/RCTMGLLocationModule.java (hunk 6 above)
Comment (ferdicus) [reply android/rctmgl/src/main/java/com/mapbox/rctmgl/modules/RCTMGLLocationModule.java]: resolved via [this](https://github.com/react-native-mapbox-gl/maps/pull/546/commits/bc97101b4dbc59c3a4ec9c21e74150d0ae9ba427)

**Comment by ferdicus** [comment]
@RichardLindhout: I reset some commits to get rid of the linting stuff in order to make it easier to reason about the changes that are actually relevant for this PR.

@kristfal: I've updated the gitignore and also commited a `yarn.lock` file for people using `yarn` instead of `npm`

How does one trigger CI builds? 
How would I go about duplicating the CI build process locally?

Thanks

File: ios/RCTMGL/RCTMGLLocationModule.m [hunk 7]
```diff
@@ -71,6 +71,11 @@ - (void)stopObserving
     [locationManager stop];
 }
 
+RCT_EXPORT_METHOD(setMinDisplacement)
+{
+    [locationManager setMinDisplacement];
+}
+
```
Comment (ferdicus) [inline ios/RCTMGL/RCTMGLLocationModule.m]: @kristfal @mattijsf : can I get some ios expertise here please (haven't really done anything ios wise before)

On compilation in XCode I get an error for this method and `#start` on line 59, that reads:

```
No visible @interface for 'RCTMGLLocationManager' declares the selector 'start'
No visible @interface for 'RCTMGLLocationManager' declares the selector 'setMinDisplacement'
```

If I understand correctly, those need to be defined in `RCTMGL/RCTMGLLocationManager.h`
Which they are though, if you check that file in this PR.

Any help/ pointers would be very appreciated.

Thanks in advance

Comment (ferdicus) [reply ios/RCTMGL/RCTMGLLocationModule.m]: hey there, I think I solved it.
It was related to the exported methods from the module not matching the interface.
I did get it to build with the changes.

**Comment by ferdicus** [comment]
@kristfal: I think this is now ready for a review

Thanks

**Comment by ferdicus** [comment]
Maybe @mattijsf or @mfazekas have some time as well?

Would love to get this in before getting new merge conflicts 😅 

Thanks in advance

**Comment by ferdicus** [comment]
@RichardLindhout, if you find the time, could you please edit the title.

Maybe something like: `feat(displacement): add setDisplacement functionality`
Or something similar. Just to signal, that this is no longer a `WIP`

Thanks in advance

**Comment by ferdicus** [comment]
@kristfal, really hate to be so annoying with this :(
Can you give me an eta or maybe I can help with anything to get the ball rolling on this?

Thanks in advance and best start into the new year! 🎆

**Comment by RichardLindhout** [comment]
> @RichardLindhout, if you find the time, could you please edit the title.
> 
> Maybe something like: `feat(displacement): add setDisplacement functionality`
> Or something similar. Just to signal, that this is no longer a `WIP`
> 
> Thanks in advance

Thanks for your work!! I'm a little busy lately with some deadlines.

**Comment by ferdicus** [comment]
> Thanks for your work!! I'm a little busy lately with some deadlines.

No need to apologize - thanks for doing most of the work and renaming 👍

**Comment by kristfal** [comment]
Hey, looks good and approved. Sorry for the slow review, Christmas holidays and all that :) 

You able to merge, or want me to do it?

**Comment by ferdicus** [comment]
> Hey, looks good and approved. Sorry for the slow review, Christmas holidays and all that :)
> 
> You able to merge, or want me to do it?

Hey 👋 

Thanks, no worries - I assumed as much, hope you had some relaxing days.

I'm right in the middle of adding an example as well.

Afterwards feel free to merge - I do not have any rights.

PS: If it's ok for you I would request some rights to take care of the `depfu` PRs.
Feel like they're bogging you down and I wouldn't have an issue looking after them once in a while.

**Comment by ferdicus** [comment]
ok, @kristfal - I'm done :)

**Comment by ferdicus** [comment]
@kristfal, still good to go :)

Thanks in advance

**Comment by ferdicus** [comment]
@kristfal , sorry to be a pain in the butt, but *ping*

Also, how difficult would it be to release a new version afterwards?


Again, if I can help with any of the depfu PRs - would gladly have a look.

Thanks in advance.
